From f909646ec325e4894c5e0e12bc1b436ff38be2ff Mon Sep 17 00:00:00 2001
From: Gerald Schaefer <gerald....@de.ibm.com>
Date: Wed, 20 Sep 2017 18:09:45 +0200
Subject: [PATCH 3/3] task-size-overrun: fix problem with dynamic pagetable
 upgrade on s390x (again)

Commit a06eeed7e005 ("task-size-overrun: fix problem with dynamic pagetable
upgrade on s390x") fixed a problem with task size detection and the dynamic
pagetable upgrade from 3 to 4 levels on s390.

Kernel 4.13 now introduced support for 5 level page tables on s390, which
results in task-size-overrun trying to loop from 2^53 (4 level task size)
up to 0xfffffffffffff000 (-4096, 5 level task size).

Fix this again by increasing addr in the loop for s390x as soon as we
exceed the 2^53 limit.

Signed-off-by: Gerald Schaefer <gerald....@de.ibm.com>
---
 tests/task-size-overrun.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/tests/task-size-overrun.c b/tests/task-size-overrun.c
index 87a9dfc..0232396 100644
--- a/tests/task-size-overrun.c
+++ b/tests/task-size-overrun.c
@@ -107,6 +107,8 @@ static unsigned long find_task_size(void)
 #if defined(__s390x__)
 		if (addr > (1UL << 42) && addr < (1UL << 53))
 			addr = 1UL << 53;
+		if (addr == (1UL << 53) + getpagesize())
+			addr = -4096;
 #endif
 	}
 	/* addr wrapped around */
-- 
2.7.4

